### ISSUE TITLE
NVDA does not announce math structure and role inside a contenteditable

### ISSUE THREAD

**Issue by mitchellevan** [nvaccess/nvda#16035]
### Steps to reproduce:

Before testing, installed MathPlayer 4 as instructed in [NVDA documentation](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#ReadingMath).

#### HTML `<math>` containing MathML

1. Go to [MathML Examples (Freedom Scientific)](https://support.freedomscientific.com/content/html/jawshq/MathML-Samples.html)
2. Use NVDA arrow keys to locate the "Quadratic Formula" math element. (Not inside an editor.)
3. Go to [MathML Examples in contenteditable (CodePen)](https://codepen.io/TPG/full/zYboGOm/7a48f9b74c541f2706d686ceacdae628)
4. Move focus into the editor.
5. Use NVDA arrow keys to locate the "Quadratic Formula" math element inside the editor.

#### ARIA `role="math"`

1. Go to [Math inside a rich text editor - testing user agent support (CodePen)](https://codepen.io/TPG/full/GRrKdop)
2. Use NVDA arrow keys to locate a math element not inside the editor.
3. Under "Enter some text and math", move focus into the editor.
4. Use NVDA arrow keys to locate the math element inside the editor.

#### HTML `<math>` containing MathML

Outside of a `contenteditable` NVDA announces an HTML MathML element as expected. Speech Viewer example (Firefox):

```
heading    level 2  Quadratic Formula
down arrow
  x  equals   the fraction with numerator   negative   b  plus or minus   the square root of   b  squared   minus   4       a      c     and denominator   2       a   
```

However, inside of a `contenteditable` NVDA announces the text of the HTML MathML element but not its structure, and NVDA does not announce the fact that it is a math element. Speech Viewer example (Firefox):

```
out of list  Edit some math:
down arrow
section  editable    heading    level 2  Area of a circle
enter
Edit some math:  section  editable  heading    level 2  Area of a circle
down arrow
  π
down arrow
heading    level 2  Quadratic Equation
down arrow
  x
down arrow
heading    level 2  Quadratic Formula
down arrow
  x
right arrow
equals
right arrow
equals
right arrow
dash
right arrow
dash
right arrow
b
right arrow
b
right arrow
plus or Minus
right arrow
plus or Minus
right arrow
b
right arrow
b
right arrow
right arrow
2
right arrow
right arrow
dash
right arrow
right arrow
4
right arrow
right arrow
right arrow
a
right arrow
right arrow
⁢
right arrow
⁢
right arrow
c
right arrow
c
right arrow
2
right arrow
2
right arrow
⁢
right arrow
⁢
right arrow
a
right arrow
a
right arrow
blank
```

The above problem is the same in Firefox and Chrome.

#### ARIA `role="math"`

There is no problem in Firefox.

In Chrome, outside of a `contenteditable`: NVDA announces the math content as expected. (It does not announce the fact that it is a math element; this might or might not be a problem in this context.)

In Chrome, inside of a `contenteditable`: NVDA announces neither the math content nor the fact that it is a math element. Instead it announces nothing, not even "blank".

### Expected behavior:

Outside of a `contenteditable` NVDA should announce the math structure and content, and optionally the fact that it is a math element.

Inside of a `contenteditable` NVDA should announce the math structure and content, and the fact that it is a math element.
- Announcing the math structure and content allows a user to understand the math together with its surrounding content.
- Informing the user of the fact that it is a math element prompts the user to know what kind of mechanism they should use for editing it.

### NVDA logs, crash dumps and other attachments:

none

#### NVDA installed/portable/running from source:

NVDA downloaded from NVAccess.org and installed

#### NVDA version:

Version: 2023.2 (2023.2.0.29051)

#### Windows version:

Windows 10

#### Name and version of other software in use when reproducing the issue:

Chrome 120
Firefox 121.0.1

#### Other information about your system:

none

#### Does the issue still occur after restarting your computer?

yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

no

#### If NVDA add-ons are disabled, is your problem still occurring?

n/a

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?

not attempted

**Comment by mitchellevan** [comment]
Similar bug in JAWS: https://github.com/FreedomScientific/standards-support/issues/799

**Comment by mitchellevan** [comment]
Chrome and Firefox both report MathML element semantics in devtools accessibility. That said...

If Chrome is partially responsible, please update [Chrome Issue 1245715: JAWS not recognizing role=math element inside contentEditable](https://bugs.chromium.org/p/chromium/issues/detail?id=1245715)

If Firefox is partially responsible, please file a bug for it.

**Comment by Adriani90** [comment]
cc: @jcsteh

**Comment by Adriani90** [comment]
@mitchellevan mathPlayer is mainly discontinued and replaced by Mathcat. Can you reproduce this issue with MathCat as well?
cc: @NSoiffer

**Comment by Adriani90** [comment]
cc: @SaschaCowley since you improved NVDA's behavior in contentEditable, maybe you could look into this as well?

**Comment by Adriani90** [comment]
1. In Chrome the coresponding issue has been fixed in version 95
2. In content editable, while in focus mode, Mathcat does not work. This works only in browse mode. It is probably due to the fact that mathcat is designed for reading math, but not for editing math. So editing math is another piece of work, not sure whether this is possible in a browser at all in an accessible way. This probably needs a separate UI. @NSoiffer please correct me if I am wrong.
3. Reading the role of math is in my opinion not required, it just introduces additional verbosity. users can already understand it is math by following ways
* Use left and right arrow in browse mode, the whole math expression is being read before entering the it with the enter key.
* Users already know the context while reading, so given the fact that NVDA reads the letters, numbers and symbols in a math expression users would know that this is an expression.
* When an user enters a math expression from browse mode, currently after integration of math the user gets an indication by speech or sound when the math expresison begins and when it ends.

**Comment by NSoiffer** [comment]
Yes, editing math is a completely different beast. There are some accessible math editors out there. It is _not_ something I intend to work on with MathCAT.

I also agree that there is no reason to announce "math". If people want to know it is math and they are using MathCAT, they can turn on the "beep-boop" sounds. In general, IMHO, something is either obviously math or it doesn't matter whether it is math or text (e.g., a < 2). For braille, it likely matters as many braille codes have indicators for the start and end of math, but that is independent of speech.
